js/src/common/components/Alert.tsx

```diff
@@ -1,31 +1,33 @@
-import Component from '../Component';
+import Component, { ComponentAttrs } from '../Component';
 import Button from './Button';
 import listItems from '../helpers/listItems';
 import extract from '../utils/extract';
+import Mithril from 'mithril';
+
+export interface AlertAttrs extends ComponentAttrs {
+  /** The type of alert this is. Will be used to give the alert a class name of `Alert--{type}`. */
+  type?: string;
+  /** An array of controls to show in the alert. */
+  controls?: Mithril.Children;
+  /** Whether or not the alert can be dismissed. */
+  dismissible?: boolean;
+  /** A callback to run when the alert is dismissed */
+  ondismiss?: Function;
+}
 
 /**
  * The `Alert` component represents an alert box, which contains a message,
  * some controls, and may be dismissible.
- *
- * ### Attrs
- *
- * - `type` The type of alert this is. Will be used to give the alert a class
- *   name of `Alert--{type}`.
- * - `controls` An array of controls to show in the alert.
- * - `dismissible` Whether or not the alert can be dismissed.
- * - `ondismiss` A callback to run when the alert is dismissed.
- *
- * All other attrs will be assigned as attributes on the DOM element.
  */
-export default class Alert extends Component {
-  view(vnode) {
+export default class Alert<T extends AlertAttrs = AlertAttrs> extends Component<T> {
```

Comment:
This is probably just me not knowing TypeScript well enough, but would this work as well?

```suggestion
export default class Alert extends Component<AlertAttrs> {
```

Reply:
Especially seeing that `T` is not used anywhere within the class.

Reply:
This would work until a class tries to extend Alert and pass its own ComponentAttrs.

Reply:
Hmm, okay, but even the base interfaces are extensible with any custom keys, right? But I see the point...

Reply:
> Hmm, okay, but even the base interfaces are extensible with any custom keys, right? But I see the point...

Not sure what you mean by this. We can pass things to `Component<>` because we have basically the same `<T extends ... = ..>` but with `ComponentAttrs`. We need to explicitly allow this in Component classes, else if they pass something it won't be registered as component props because we're already passing e.g. AlertAttrs ``and that can't be changed.

Reply:
I meant that `ComponentAttrs` (and therefore `AlertAttrs`) already allows any key in attrs, beyond the well-defined ones. So the usefulness of custom typings for the subclasses' props is limited.

Reply:
The usefulness is to type the props themselves so when you use the class it tells you what the props should be & their type, and then when accessing them from within the class too.